### PR TITLE
Feat/reputation oracle v0

### PR DESCRIPTION
**Pull request by eugenvoronov**
## Description
Develop basic Reputation Oracle Server functionality to interact with Exchange Oracle and Recording Oracle to check their results, write reputation information to a new file and database.

## Summary of changes
- [x] Implement SDK
- [x] Create new database architecture base on [documentation](https://www.notion.so/human-protocol/Reputation-Oracle-9c56f7fbc52e4f62a7c443878cd576d8?pvs=4#aa81d986f44944f6a01453317fa962e3)
- [x] Create Webhook module
- [x] Create Reputation module  
- [x] Create cron job: [Incoming webhooks processing](https://www.notion.so/human-protocol/Reputation-Oracle-9c56f7fbc52e4f62a7c443878cd576d8?pvs=4#82258851071c4785a5692e43956eba68)
- [x]  Create cron job: [Reputation processing](https://www.notion.so/human-protocol/Reputation-Oracle-9c56f7fbc52e4f62a7c443878cd576d8?pvs=4#865337bd5d5e4cccbe02a01db0a980cd)
- [x] Refactored config env vars
- [x] Implemented unit tests

## Related issues
[Reputation Oracle Server#506](https://github.com/humanprotocol/human-protocol/issues/506)
[[Reputation Oracle] Unit tests#573](https://github.com/humanprotocol/human-protocol/issues/573)

### CODE REVIEW

File: packages/examples/fortune-v2/reputation-oracle/server/src/main.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+      
+    }),
+  );
+  app.use(json({ limit: "5mb" }));
+  app.use(urlencoded({ limit: "5mb", extended: true }));
+
+  const config = new DocumentBuilder()
+    .addBearerAuth()
+    .setTitle("Fortune Exchange Oracle API")
+    .setDescription("Swagger Fortune Exchange Oracle API")
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/src/main.ts:49]:
```suggestion
    .setTitle("Fortune Reputation Oracle API")
    .setDescription("Swagger Fortune Reputation Oracle API")
```

File: packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.dto.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+export class WebhookIncomingCreateDto {
+  @ApiProperty()
+  @IsEnum(ChainId)
+  public chainId: ChainId;
+
+  @ApiProperty()
+  @IsString()
+  public escrowAddress: string;
+
+  public status?: WebhookStatus;
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.dto.ts:17]: not necessary

File: packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.dto.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  public waitUntil?: Date;
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.dto.ts:18]: not necessary, the dto doesn't need it, if you need it for something else maybe you can extend

Comment (eugenvoronov) [reply packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.dto.ts:18]: We are using this properties for createIncomingWebhook(dto: WebhookIncomingCreateDto): Promise<boolean> {} 
Before, for this reason we had used interfaces. 
> maybe you can extend
What do you mean here?

Comment (portuu3) [reply packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.dto.ts:17]: Status and waitUntil ar not needed properties for the object you receive in the controller. So why to include them in the dto?
If you need to a have a class/dto or whatever with the other 2 properties, you can extend from this dto, but don't add properties you dont need, actually in this case you remove status and waitUntil and nothing would happen

File: packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.service.ts [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+          id: webhookEntity.id,
+        }, { 
+          resultsUrl: finalResultsUrl,
+          checkPassed,
+          status: WebhookStatus.PAID,
+          retriesCount: 0
+        })
+      
+      const recipients = finalResults.map(item => item.workerAddress);
+      const amounts = new Array(recipients.length).fill(BigNumber.from(manifest.fundAmount).div(recipients.length));
```
Comment (eugenvoronov) [inline packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook.service.ts:184]: Please, take a look here, we are trying to get the payout amount for each worker from the fund amount, but now it includes fee amount (fundAmount + fee), I suggest separating these 2 fields and storing them separately in the manifest. Or other suggestions? @portuu3 

https://github.com/humanprotocol/human-protocol/pull/520/files#r1228372175

File: packages/examples/fortune-v2/reputation-oracle/server/src/common/decorators/webhook.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  MOONBASE_ALPHA = 1287,
+  LOCALHOST = 1338,
+}
+
+export enum WebhookStatus {
+  PENDING = 'PENDING',
+  COMPLETED = 'COMPLETED',
+  FAILED = 'FAILED',
+  PAID = 'PAID',
+}
```
[the remark this answers is not in the document]
Comment (flopez7) [reply packages/examples/fortune-v2/reputation-oracle/server/src/common/decorators/webhook.ts:27]: Maybe the ChainId one can be imported from SDK 
`import { ChainId } from '@human-protocol/sdk';`

File: packages/examples/fortune-v2/reputation-oracle/server/src/modules/reputation/reputation.entity.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import { Column, Entity } from 'typeorm';
+
+import { NS } from '../../common/constants';
+import { BaseEntity } from '../../database/base.entity';
+import { ReputationEntityType } from '../../common/decorators';
+
+@Entity({ schema: NS, name: 'reputation' })
+export class ReputationEntity extends BaseEntity {
+  @Column({ type: 'int' })
+  public chainId: number;
```
Comment (flopez7) [inline packages/examples/fortune-v2/reputation-oracle/server/src/modules/reputation/reputation.entity.ts:10]: Shouldn't it use the enum?
```suggestion
  public chainId: ChainId;
```

File: packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook-incoming.entity.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { Column, Entity } from 'typeorm';
+
+import { NS } from '../../common/constants';
+import { BaseEntity } from '../../database/base.entity';
+import { WebhookStatus } from '../../common/decorators';
+
+@Entity({ schema: NS, name: 'webhook_incoming' })
+export class WebhookIncomingEntity extends BaseEntity {
+  @Column({ type: 'int' })
+  public chainId: number;
```
Comment (flopez7) [inline packages/examples/fortune-v2/reputation-oracle/server/src/modules/webhook/webhook-incoming.entity.ts:10]:
```suggestion
  public chainId: ChainId;
```

File: packages/sdk/typescript/human-protocol-sdk/src/interfaces.ts
Comment (flopez7) [inline packages/sdk/typescript/human-protocol-sdk/src/interfaces.ts:1]: Why did you add this? It is not used anywhere

File: packages/sdk/typescript/human-protocol-sdk/src/escrow.ts
Comment (flopez7) [inline packages/sdk/typescript/human-protocol-sdk/src/escrow.ts:1]: You should remove this change. It was fine before. With that change you are modifying the correct order

`function setup(
        address _reputationOracle,
        address _recordingOracle,
        uint8 _reputationOracleFeePercentage,
        uint8 _recordingOracleFeePercentage,
        string memory _url,
        string memory _hash
    )`

File: packages/examples/fortune-v2/reputation-oracle/server/src/app.module.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+      useClass: HttpValidationPipe,
+    },
+  ],
+  imports: [
+    ScheduleModule.forRoot(),
+    ConfigModule.forRoot({
+      envFilePath: process.env.NODE_ENV
+        ? `.env.${process.env.NODE_ENV as string}`
+        : '.env',
+    }),
```
Comment (flopez7) [inline packages/examples/fortune-v2/reputation-oracle/server/src/app.module.ts:26]:
```suggestion
    ConfigModule.forRoot({
      envFilePath: process.env.NODE_ENV
        ? `.env.${process.env.NODE_ENV as string}`
        : '.env',
      validationSchema: envValidator,
    }),
```

**Review by flopez7 — CHANGES_REQUESTED** [review]

File: packages/examples/fortune-v2/reputation-oracle/server/src/common/test/constants.ts
[the remark this answers is not in the document]
Comment (eugenvoronov) [reply packages/examples/fortune-v2/reputation-oracle/server/src/common/test/constants.ts:1]: Perhaps we can do this, but I suggest leaving it separate

File: packages/examples/fortune-v2/reputation-oracle/server/.env.example [hunk 9]
```diff
@@ -0,0 +1,29 @@
+# General
+NODE_ENV=
+HOST=
+PORT=
+FE_URL=
+SESSION_SECRET=
+PASSWORD_SECRET='$2b$10$EICgM2wYixoJisgqckU9gu'
```
Comment (flopez7) [inline packages/examples/fortune-v2/reputation-oracle/server/.env.example:7]: You should remove this password

File: packages/examples/fortune-v2/reputation-oracle/server/test/constants.ts
Comment (flopez7) [inline packages/examples/fortune-v2/reputation-oracle/server/test/constants.ts:1]: This file is duplicated in `packages/examples/fortune-v2/reputation-oracle/server/src/common/test/constants.ts`

**Review by flopez7 — CHANGES_REQUESTED** [review]

**Review by flopez7 — APPROVED** [review]

File: packages/examples/fortune-v2/reputation-oracle/server/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+      - POSTGRES_PASSWORD=qwerty
+      - POSTGRES_DB=reputation-oracle
+      - POSTGRES_PORT=5435
+      - POSTGRES_SYNC=false
+    logging:
+      options:
+        max-size: 10m
+        max-file: "3"
+    ports:
+      - '5435:5432'
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/docker-compose.yml:19]:
```suggestion
      - '5432:5432'
```

File: packages/examples/fortune-v2/reputation-oracle/server/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+services:
+  postgres:
+    image: postgres:latest
+    restart: always
+    environment:
+      - POSTGRES_HOST=0.0.0.0
+      - POSTGRES_USER=operator
+      - POSTGRES_PASSWORD=qwerty
+      - POSTGRES_DB=reputation-oracle
+      - POSTGRES_PORT=5435
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/docker-compose.yml:12]:
```suggestion
      - POSTGRES_PORT=5432
```

File: packages/examples/fortune-v2/reputation-oracle/server/package.json [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    "start:dev": "nest start --watch",
+    "start:debug": "nest start --debug --watch",
+    "start:prod": "node dist/main",
+    "migration:create": "typeorm-ts-node-commonjs migration:create",
+    "migration:generate": "typeorm-ts-node-commonjs migration:generate -d typeorm.config.ts",
+    "migration:revert": "typeorm-ts-node-commonjs migration:revert -d typeorm.config.ts",
+    "migration:run": "typeorm-ts-node-commonjs migration:run -d typeorm.config.ts",
+    "migration:show": "typeorm-ts-node-commonjs migration:show -d typeorm.config.ts",
+    "docker:db:up": "docker-compose up -d && yarn migration:run",
+    "docker:db:down": "docker-compose down",
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/package.json:21]:
```suggestion
    "docker:db:up": "docker compose up -d && yarn migration:run",
    "docker:db:down": "docker compose down",
```

File: packages/examples/fortune-v2/reputation-oracle/server/src/common/config/env.ts [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+  PORT: Joi.string().default(5000),
+  FE_URL: Joi.string().default('http://localhost:3001'),
+  SESSION_SECRET: Joi.string().default('session_key'),
+  PASSWORD_SECRET: Joi.string().default('$2b$10$EICgM2wYixoJisgqckU9gu'),
+  // Database
+  DB_TYPE: Joi.string().default('postgres'),
+  POSTGRES_HOST: Joi.string().default('127.0.0.1'),
+  POSTGRES_USER: Joi.string().default('operator'),
+  POSTGRES_PASSWORD: Joi.string().default('qwerty'),
+  POSTGRES_DB: Joi.string().default('job-launcher'),
```
Comment (portuu3) [inline packages/examples/fortune-v2/reputation-oracle/server/src/common/config/env.ts:47]:
```suggestion
  POSTGRES_DB: Joi.string().default('reputation-oracle'),
```

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/examples/fortune-v2/reputation-oracle/server/package.json (hunk 12 above)
Comment (eugenvoronov) [reply packages/examples/fortune-v2/reputation-oracle/server/package.json:21]: https://docs.docker.com/compose/migrate/ in my machine no difference what kind of signature to use, lets update it

**Review by portuu3 — APPROVED** [review]
